RSA Key management and new JWK endpoint

marcospri opened the pull request:
Requires:

 * https://github.com/hypothesis/lms/pull/3876

## Testing

- Run the migration to create the new table 

```hdev alembic upgrade head```

- Create a couple of keys in `make shell`

```
from lms.services import RSAKeyService

rsa_service = request.find_service(RSAKeyService)

key_1 = rsa_service.generate()

key_2 = rsa_service.generate()
tm.commit()
```

- Check the new endpoint exposing them: http://localhost:8001/lti/1.3/jwks
  

- Mark one of the as expired in a new `make shell`

```
key = session.query(models.RSAKey).first()
key.expired=True
tm.commit()
```


- Check http://localhost:8001/lti/1.3/jwks only exposes one key now.

marcospri commented on conf/supervisord-dev.conf:
```diff
@@ -3,7 +3,7 @@ nodaemon = true
 
 [program:web]
-command=gunicorn --paste conf/development.ini
+command=gunicorn --workers 1 --paste conf/development.ini
```
Making the default explicit, this way is easier to change it in dev. 

I haven't found a good way to do it in the .ini file.

marcospri commented on lms/models/key.py:
```diff
@@ -0,0 +1,39 @@
+    aes_cipher_iv = sa.Column(sa.LargeBinary)
+
+    public_key = sa.Column(sa.Unicode)
+    _private_key = sa.Column(sa.LargeBinary)
```
I probably should rename this to something indicating is "aes" encoded.

marcospri commented on lms/services/lti_outcomes.py:
```diff
@@ -1,26 +1,83 @@
-    def __init__(self, _context, request):
-        self.oauth1_service = request.find_service(name="oauth1")
-        self.http_service = request.find_service(name="http")
+    def __init__(  # pylint:disable=too-many-arguments
```
Changed this service to get the services injected and use a factory.

It's just how we tend to to thing now and allows to create a the service from scratch if necessary (for example in `make shell` which was necessary for the certification testing)

marcospri commented on lms/services/ltia_http.py:
```diff
@@ -0,0 +1,93 @@
+        self._application_instance = application_instance
+        self._jwt_service = jwt_service
+        self._key_service = key_service
+        self._aes_service = aes_service
```
Not used

marcospri commented on lms/services/ltia_http.py:
```diff
@@ -0,0 +1,93 @@
+        https://datatracker.ietf.org/doc/html/rfc7523
+        https://canvas.instructure.com/doc/api/file.oauth_endpoints.html#post-login-oauth2-token
+
+        Scopes:
```
Remove comments

marcospri commented on lms/services/user.py:
```diff
@@ -17,8 +17,7 @@ class UserService:
 
-    def __init__(self, application_instance_service, db, h_authority: str):
-        self._application_instance_service = application_instance_service
+    def __init__(self, db, h_authority: str):
```
Not used :man_shrugging:

marcospri commented on requirements/requirements.in:
```diff
@@ -12,6 +12,7 @@ oauthlib
 psycopg2
 pycryptodomex
 PyJWT[crypto]
+python-jose[crypto]
```
After we move the reminding JWT bits to the service we could see if its possible to settle only on one JWT library.

jon-betts replied:
What's wrong with the existing one? I suspect adding a new JWT library would get the Sean upside-down smile of approval.

jon-betts commented on lms/models/key.py:
```diff
@@ -0,0 +1,41 @@
+from lms.models import CreatedUpdatedMixin
+
+
+class Key(CreatedUpdatedMixin, BASE):
```
Is this a `JSONWebKey`? Key alone is quite vague.

marcospri replied:
Went for RSAKey

jon-betts commented on lms/models/key.py:
```diff
@@ -0,0 +1,41 @@
+    public_key = sa.Column(sa.Unicode)
+    """Public key a json JWK string"""
+
+    _private_key = sa.Column(sa.LargeBinary)
```
This isn't private in a scoping sense, just in an RSA key pair sense?

```suggestion
    private_key = sa.Column(sa.LargeBinary)
```

jon-betts commented on lms/services/key.py:
```diff
@@ -0,0 +1,63 @@
+from lms.services import AESService
+
+
+class KeyService:
```
Is this the `JSONWebKeyService`?

marcospri replied:
Settled on RSAKeyService

jon-betts commented on lms/services/key.py:
```diff
@@ -0,0 +1,63 @@
+            aes_cipher_iv=aes_iv,
+        )
+        self._db.add(key)
+        return key
```
This is quite a lot in one go. It might help readability to split this into:

* `def _encrypt_private_key_as_pem(self, rsa_key, aes_iv)`
* `def _encrypt_public_key_as_jwk(self, rsa_key)`

Or something to break this up a bit

jon-betts commented on lms/services/key.py:
```diff
@@ -0,0 +1,63 @@
+
+
+class KeyService:
+    def __init__(self, db, aes_service):
+        self._db = db
+        self._aes = aes_service
```
```suggestion
    def __init__(self, db, aes_service: AESService):
        self._db = db
        self._aes_service = aes_service
```

There are quite a few crypto libraries floating about. It's nice to point out this is one of ours. This also matches the naming used elsewhere in services.

jon-betts commented on lms/services/ltia_http.py:
```diff
@@ -0,0 +1,87 @@
+            self._key_service.private_key(key),
+            payload=payload,
+            headers=headers,
+        )
```
It seems like a lot of this could be inside the `JWTService`

jon-betts commented on tests/functional/lti_certification/v13/grading/test_assignment_and_grading.py:
```diff
@@ -0,0 +1,61 @@
+
+    After that launch it is the responsibility of the Tool alone to work with the AGS API to create lineitems and scores in the Certification Suite.
+    All interaction with the Gradebook simulated by the Certification Suite can be viewed on the Results page
+    """
```
Some mega long lines in this file. Could we keep it to the PEP-8 79 chars?

jon-betts reviewed (commented):
I've put some arbitrary musings on here. Nothing very structured.

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+
+    def __init__(self, db, aes_service):
+        self._db = db
+        self._aes_service = aes_service
```
This looks like a good candidate for a `@dataclass`

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+        ).to_dict()
+        json_publick_key["use"] = "sig"
+        return json_publick_key
+
```
We have some public methods after private ones here

jon-betts commented on lms/views/lti/jwk.py:
```diff
@@ -0,0 +1,17 @@
+from lms.services import RSAKeyService
+
+
+@view_config(
+    route_name="lti.jwks",
+    request_method="GET",
+    renderer="json",
+)
+def jwks(request):
+    key_service = request.find_service(RSAKeyService)
+    keys = key_service.all()
+
+    return {"keys": [json.loads(k.public_key) for k in keys]}
```
```suggestion
@view_config(route_name="lti.jwks", request_method="GET", renderer="json")
def jwks(request):
    keys = request.find_service(RSAKeyService).all()

    return {"keys": [json.loads(key.public_key) for key in keys]}
```

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PublicFormat.SubjectPublicKeyInfo,
+        )
+        json_publick_key = jwk.RSAKey(
+            algorithm=constants.Algorithms.RS256, key=pem_public_key.decode("utf-8")
+        ).to_dict()
+        json_publick_key["use"] = "sig"
+        return json_publick_key
```
```suggestion
        json_public_key = jwk.RSAKey(
            algorithm=constants.Algorithms.RS256, key=pem_public_key.decode("utf-8")
        ).to_dict()
        json_public_key["use"] = "sig"

        return json_public_key
```

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+        """Decrypt the AES encrypted private key of `key`."""
+        return self._aes_service.decrypt(key.aes_cipher_iv, key.private_key)
+
+    def all(self):
```
This is a pretty much impossible thing to grep for. Would this be better?

```suggestion
    def get_all_keys(self):
```

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+        """Get all non-expired keys."""
+        return self._db.query(RSAKey).filter_by(expired=False).all()
+
+    def one(self):
```
```suggestion
    def get_random_key(self):
```

I also can't see any consumers for this?

jon-betts commented on tests/unit/lms/services/rsa_key_service_test.py:
```diff
@@ -0,0 +1,100 @@
+    def test_all(self, svc):
+        keys = svc.all()
+
+        assert not any(k.expired for k in keys)
```
```suggestion
        assert not any(key.expired for key in keys)
```

jon-betts replied:
It feels a bit like we should be checking that we get the ones we expect here. For example if this returned only one, this test wouldn't fail.

I like the fixture adding some noise elements, but I think it might work just as well if it returned the non expired keys and we compared to them.

jon-betts commented on tests/unit/lms/services/rsa_key_service_test.py:
```diff
@@ -0,0 +1,100 @@
+    def test_one(self, svc):
+        key = svc.one()
+
+        assert not key.expired
```
Kind of similar here, but less important. It would be nice to say `assert key in valid_keys` or something.

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,78 @@
+            public_exponent=65537, key_size=2048, backend=default_backend()
+        )
+
+        # Generate a IV for the private key AES encryption
```
```suggestion
        # Generate an IV for the private key AES encryption
```

jon-betts commented on tests/unit/lms/services/rsa_key_service_test.py:
```diff
@@ -0,0 +1,100 @@
+
+    @pytest.fixture
+    def jwk(self, patch):
+        return patch("lms.services.rsa_key.jwk")
```
Everytime I see this my brain reads it out like "I've got a good jwk: A llama, hobbit and smelter walk into a bar... "

jon-betts reviewed (commented):
I don't have a full review on this as I have a few questions.

Should the migration be in a separate PR? I'm not sure if this is safe to deploy or not.

There's no consumer for `one()`, which makes it hard to review, but one possibility that occurred to me is that instead of `all()` and `one()` we really want `get_all_public_keys()` and `get_random_public_key()` which would also perform the JSON decoding for you.

This would strongly depend on whether there are any consumers which want the raw key object or not. But if it never happens it might be nice to fess up that that's what these are for.

There are various other little suggestions and so.

marcospri commented:
> I don't have a full review on this as I have a few questions.
> 
> Should the migration be in a separate PR? I'm not sure if this is safe to deploy or not.

I think as we don't use any of this in any existing code we should we fine but to play it safe I marked the PR as `do-not-merge` and will move the migration to another PR if all looks good.


> There's no consumer for `one()`, which makes it hard to review, but one possibility that occurred to me is that instead of `all()` and `one()` we really want `get_all_public_keys()` and `get_random_public_key()` which would also perform the JSON decoding for you.
> 
> This would strongly depend on whether there are any consumers which want the raw key object or not. But if it never happens it might be nice to fess up that that's what these are for.

For the .one we'll do need the whole object (or at least the private part instead of the public one). In any case, as we don't use it in this PR I left it out for the next one. Renamed `all` to `get_all_public_jwks` and moved the json loading and some parameters needed for the jwk there.


> I'm sure there's a good reason we don't use the existing JWT library, but I'd be interested to know what it is.

Basically the current jwt lib doesn't provide a method to format PEMs as jwks but the new one does (the `jwk.RSAKey` call). The interfaces are basically the same for the rest of our usage so I think it should be pretty easy to settle in the newer one which is based on the current one.

```This library was originally based heavily on the work of the folks over at PyJWT.```

marcospri commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,74 @@
+    def get_all_public_jwks(self):
+        """Get all non-expired keys."""
+        keys = self._db.query(RSAKey).filter_by(expired=False).all()
+        jwks = []
```
I had a generator here but make the view code more awkward.

jon-betts replied:
How so? I would think this would work:

```python
@view_config(route_name="lti.jwks", request_method="GET", renderer="json")
def jwks(request):
    return {"keys": list(request.find_service(RSAKeyService).get_all_public_jwks())}
```

jon-betts commented on lms/views/lti/jwk.py:
```diff
@@ -0,0 +1,8 @@
+
+
+@view_config(route_name="lti.jwks", request_method="GET", renderer="json")
+def jwks(request):
```
We should have a doc string here describing what purpose this end-point serves.

In general, we should document everything, rather than it being an exception.

jon-betts commented on tests/unit/lms/services/rsa_key_service_test.py:
```diff
@@ -0,0 +1,95 @@
+    def test_get_all_public_jwks(self, svc, valid_keys):
+        keys = svc.get_all_public_jwks()
+
+        assert len(keys) == len(valid_keys)
+        assert set(key["kid"] for key in keys) == set(key.kid for key in valid_keys)
+        for key in keys:
+            assert key["use"] == "sig"
```
```suggestion
        expected_keys = [
            Any.dict.containing({"use": "sig", "kid": key["kid"]}) for key in keys
        ]
        assert keys == Any.list.containing(expected_keys).only()
```

Not sure if this is any better, but it would be easier to expend it to cover the JSON-ification of the key as well.

jon-betts commented on lms/services/rsa_key.py:
```diff
@@ -0,0 +1,74 @@
+
+    def get_all_public_jwks(self):
+        """Get all non-expired keys."""
+        keys = self._db.query(RSAKey).filter_by(expired=False).all()
+        jwks = []
+        for key in keys:
+            public_key = json.loads(key.public_key)
+            public_key["use"] = "sig"
+            public_key["kid"] = key.kid
+            jwks.append(public_key)
+
+        return jwks
```
Not sure if you want to be this guy, but just in case you do...

```suggestion
        return [
            dict(json.loads(key.public_key), use="sig", kid=key.kid)
            for key in self._db.query(RSAKey).filter_by(expired=False).all()
        ]
```

marcospri replied:
> Not sure if you want to be this guy, but just in case you do...
You know I do

jon-betts commented on tests/unit/lms/services/rsa_key_service_test.py:
```diff
@@ -0,0 +1,95 @@
+    def valid_keys(self):
+        # Create some noise with some expired keys.
+        factories.RSAKey.create_batch(size=3, expired=True)
+        return factories.RSAKey.create_batch(size=3, expired=False)
```
If we added some content here...

```suggestion
        return factories.RSAKey.create_batch(
            size=3, expired=False, public_key='{"key": "value"}'
        )
```

We could simplify the test above and increase the coverage:

```python
        expected_keys = [
            {"use": "sig", "kid": key["kid"], "key": "value"} for key in keys
        ]
        assert keys == Any.list.containing(expected_keys).only()
```

jon-betts reviewed (changes requested):
I think we should split the migration away from the code. This is a purely fear based response, but I'm concerned that even SQLAlchemy scanning the code could cause it to discover the missing table. This won't show in dev, but it would show in QA. In any case, I don't see much advantage in running the risk.

The other request is that we document the view that's been created, and explain why it's there rather than something like "Get JWKS" etc. `h` is full of views which no doubt made sense when written but are kind of hard to know what they are for without a lot of investigation.

As a side note, I think we should turn on checking for missing doc strings. It's gone from being a case of a few annoying cases like constructors requiring low value docstrings, to it being about 50/50 whether we document at all.

There are a few other style suggestions, which are dealers choice.
